### PR TITLE
Refactor loading of tracks from database

### PR DESCRIPTION
**Pull request by uklotzde**
Another refactoring that consolidates various functions for loading tracks from the internal database:
- Use `TrackFile` instead of `QFileInfo` for track locations
- Use `TrackRef` containing either a `TrackId` or a file path (`TrackFile`) with the location
- Use differing functions names to better distinguish overloaded functions
- Fixed outdated comments

The utility classes `TrackFile` and `TrackRef` improve type-safety and ensure correct conversion between `QUrl` <-> `QFileInfo` <-> `QSting` when dealing with for *track locations*.

Not all variants are currently used, but are required for the integration of *aoide* as an external library.

Preview: The next PR will finally fix the metadata handling (import/export) for tracks and allow to support additional fields that are only stored in file tags but not (yet) in the database.

### CODE REVIEW

File: src/library/dao/trackdao.cpp [hunk 1]
```diff
@@ -110,35 +110,32 @@ void TrackDAO::finish() {
...
-TrackId TrackDAO::getTrackId(const QString& absoluteFilePath) const {
-
-    TrackId trackId;
+TrackId TrackDAO::getTrackIdByLocation(const QString& location) const {
+    if (location.isEmpty()) {
+        return TrackId();
+    }
 
     QSqlQuery query(m_database);
     query.prepare("SELECT library.id FROM library INNER JOIN track_locations ON library.location = track_locations.id WHERE track_locations.location=:location");
```
Comment (daschuer) [inline src/library/dao/trackdao.cpp]: please break this long line.

File: src/library/dao/trackdao.cpp [hunk 2]
```diff
@@ -206,9 +203,9 @@ QList<TrackId> TrackDAO::resolveTrackIds(const QList<QFileInfo>& files,
         while (query.next()) {
             trackIds.append(TrackId(query.value(idColumn)));
         }
-        DEBUG_ASSERT(trackIds.size() <= files.size());
-        if (trackIds.size() < files.size()) {
-            qDebug() << "TrackDAO::getTrackIds(): Found only" << trackIds.size() << "of" << files.size() << "tracks in library";
+        DEBUG_ASSERT(trackIds.size() <= trackFiles.size());
+        if (trackIds.size() < trackFiles.size()) {
+            qDebug() << "TrackDAO::getTrackIds(): Found only" << trackIds.size() << "of" << trackFiles.size() << "tracks in library";
```
Comment (daschuer) [inline src/library/dao/trackdao.cpp]: long line.

File: src/library/dao/trackdao.cpp [hunk 3]
```diff
@@ -1397,6 +1394,44 @@ TrackPointer TrackDAO::getTrackById(TrackId trackId) const {
...
 }
 
+TrackId TrackDAO::getTrackIdByRef(
+        const TrackRef& trackRef) const {
+    if (trackRef.getId().isValid()) {
+        return trackRef.getId();
+    }
+    {
+        GlobalTrackCacheLocker cacheLocker;
+        const auto track = cacheLocker.lookupTrackByRef(trackRef);
```
Comment (daschuer) [inline src/library/dao/trackdao.cpp]: pTrack

File: src/library/dao/trackdao.cpp [hunk 4]
```diff
@@ -1397,6 +1394,44 @@ TrackPointer TrackDAO::getTrackById(TrackId trackId) const {
...
+    if (!trackRef.isValid()) {
+        return TrackPointer();
+    }
+    {
+        GlobalTrackCacheLocker cacheLocker;
+        auto track = cacheLocker.lookupTrackByRef(trackRef);
+        if (track) {
+            return track;
+        }
+    }
```
Comment (daschuer) [inline src/library/dao/trackdao.cpp]: the code above is used in TrackDAO::getTrackIdByRef() as well. Common function?

File: src/library/dao/trackdao.cpp [hunk 5]
```diff
@@ -1962,41 +1997,53 @@ void TrackDAO::detectCoverArtForTracksWithoutCover(volatile const bool* pCancel,
...
+            qWarning()
+                    << "Failed to get track"
+                    << trackRef;
+            return TrackPointer();
         }
     }
 
+    DEBUG_ASSERT(trackRef.hasLocation());
+    addTracksPrepare();
+    const auto trackLocation = trackRef.getLocation();
```
Comment (daschuer) [inline src/library/dao/trackdao.cpp]: unused?

File: src/library/trackcollection.cpp [hunk 6]
```diff
@@ -169,19 +170,16 @@ QList<TrackId> TrackCollection::resolveTrackIdsFromUrls(
 
 QList<TrackId> TrackCollection::resolveTrackIdsFromLocations(
         const QList<QString>& locations) {
-    QList<QFileInfo> fileInfoList;
-    foreach(QString fileLocation, locations) {
-        QFileInfo fileInfo(fileLocation);
-        fileInfoList.append(fileInfo);
+    QList<TrackFile> trackFiles;
+    trackFiles.reserve(locations.size());
+    for (const QString& location : locations) {
```
Comment (daschuer) [inline src/library/trackcollection.cpp]: does qt::as_const make sense here?

File: src/util/dnd.cpp [hunk 7]
```diff
@@ -97,12 +95,12 @@ bool allowLoadToPlayer(
...
 //static
-QList<QFileInfo> DragAndDropHelper::supportedTracksFromUrls(
+QList<TrackFile> DragAndDropHelper::supportedTracksFromUrls(
         const QList<QUrl>& urls,
         bool firstOnly,
         bool acceptPlaylists) {
-    QList<QFileInfo> fileLocations;
-    for (QUrl url : urls) {
+    QList<TrackFile> trackFiles;
+    for (const QUrl& url : urls) {
```
Comment (daschuer) [inline src/util/dnd.cpp]: qt::as_const() .. unsure.

**Review by daschuer — CHANGES_REQUESTED** [review]
Thank you. I have added some minor comments. Manual test is pending.

**Comment by daschuer** [comment]
This branch works good.

File: src/library/trackcollection.cpp (hunk 6 above)
Comment (uklotzde) [reply src/library/trackcollection.cpp]: No, already a const-ref.

File: src/library/dao/trackdao.cpp (hunk 4 above)
Comment (uklotzde) [reply src/library/dao/trackdao.cpp]: No, doesn't help. There are subtle differences and you could only extract a single line. I already tried it.

File: src/util/dnd.cpp (hunk 7 above)
Comment (uklotzde) [reply src/util/dnd.cpp]: Already passed by const-ref

**Comment by uklotzde** [comment]
All issues addressed.

File: src/library/dao/trackdao.cpp (hunk 4 above)
Comment (daschuer) [reply src/library/dao/trackdao.cpp]: OK

File: src/library/trackcollection.cpp (hunk 6 above)
Comment (daschuer) [reply src/library/trackcollection.cpp]: OK

File: src/util/dnd.cpp (hunk 7 above)
Comment (daschuer) [reply src/util/dnd.cpp]: OK

**Comment by daschuer** [comment]
is still working nice. LGTM Thank you.

**Comment by uklotzde** [comment]
Hopefully I fixed all missing compatibility for qAsConst in master. The issues only became obvious on Travis with the ancient Qt version.

**Comment by uklotzde** [comment]
...caused by merging 2.2 into master, which already has been optimized to avoid unneeded #include directives.

**Pull request merged**